System/DataModule.cpp

```diff
@@ -1,764 +1,497 @@
+			// Mark the one we are about to add to the list as the Original now - this is now the actual Original Preset instance
+			pEntClone->m_IsOriginalPreset = true;
+
+			// Insert clone into the list of entries where it will now be owned
+			// New file location specified
+			if (readFromFile != "Same") {
+				m_PresetList.push_back(PresetEntry(pEntClone, readFromFile));
+				m_EntityList.push_back(pEntClone);
+
+				// If same file specified for new instance, use the one in the last entry
+			} else if (!m_PresetList.empty()) {
+				m_PresetList.push_back(PresetEntry(pEntClone, m_PresetList.back().m_FileReadFrom));
+				m_EntityList.push_back(pEntClone);
+
+			} else {
+				RTEAbort("Tried to add first entity instance to data module " + m_FileName + " without specifying a data file!");
+			}
+			// Add it to the type map and quit
+			added = AddToTypeMap(pEntClone);
+			RTEAssert(added, "Unexpected problem while adding Entity instance \"" + pEntToAdd->GetPresetName() + "\" to the type map of data module: " + m_FileName);
+		}
+		return added;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	bool DataModule::GetGroupsWithType(std::list<std::string> &groupList, std::string withType) {
+		bool foundAny = false;
+
+		if (withType == "All" || withType.empty()) {
+			// Just give all the groups
+			for (std::list<std::string>::const_iterator gItr = m_GroupRegister.begin(); gItr != m_GroupRegister.end(); ++gItr) { groupList.push_back(*gItr); }
+		} else {
+			// Go through all entities, adding the groups the ones of the right type belong to
+			// Find the type entry in the type map
+			map<std::string, std::list<std::pair<std::string, Entity *> > >::iterator clsItr = m_TypeMap.find(withType);
+			if (clsItr == m_TypeMap.end()) { return false; }
+
+			const std::list<std::string> *pGroupList = 0;
+			// Go through all the entities of that type, adding the groups they belong to
+			for (std::list<std::pair<std::string, Entity *> >::iterator instItr = clsItr->second.begin(); instItr != clsItr->second.end(); ++instItr) {
+				pGroupList = instItr->second->GetGroupList();
+				// Get the grouped entities, without transferring ownership
+				for (std::list<std::string>::const_iterator gItr = pGroupList->begin(); gItr != pGroupList->end(); ++gItr) {
+					groupList.push_back(*gItr);
+					foundAny = true;
+				}
+			}
+			// Make sure there are no dupe groups in the list
+			groupList.sort();
+			groupList.unique();
+		}
+		return foundAny;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	bool DataModule::GetAllOfGroup(std::list<Entity *> &entityList, std::string group, std::string type) {
+		if (group.empty()) { return false; }
+
+		bool foundAny = false;
+
+		// Look in all classes, so only look in the Entity level, which includes all!
+		if (type.empty() || type == "All") {
```

Comment:
Looking at it, both sides of this if-else check are basically identical. We can cut down a bunch of code by replacing the `m_TypeMap.find(x)` so x is "Entity" if type is empty or all, or type otherwise. A ternary like the following will do
`m_TypeMap.find((type.empty() || type == "All") ? "Entity" : type);`

As for the RTEAssert in the 2nd half, put it in an if or modify the condition so it'll only trigger when we run into that problem while looking for specific entity types.